docs/examples/userguide/buffer/matrix_with_buffer.py

```diff
@@ -0,0 +1,48 @@
+# distutils: language = c++
+from cython.cimports.cpython import Py_buffer
+from cython.cimports.libcpp.vector import vector
+
+@cython.cclass
+class Matrix:
+    ncols: cython.Py_ssize_t
+    shape = cython.declare(Py_ssize_t[2])
+    strides = cython.declare(Py_ssize_t[2])
```

Comment:
`shape: cython.Py_ssize_t[2]` doesn't work?

Reply:
@da-woods you are right :-/. I missed that. What I see as a pain in pure python mode is that it does not raise error that `Py_ssize_t` is not declared but it just fallbacks to object instead :-/. I expect that it should behave like regular python [1]:
```python
$ python
Python 3.9.12 (main, May  8 2022, 18:05:13)
[Clang 12.0.0 (clang-1200.0.32.29)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> aa: Spam
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'Spam' is not defined
```


Btw. it is funny that `declare()` is fine with just `Py_ssize_t` - here is part of generated c file:
```
/* "matrix_with_buffer.py":7
 * 
 * @cython.cclass
 * class Matrix:             # <<<<<<<<<<<<<<
 *     ncols: cython.Py_ssize_t
 *     shape = cython.declare(Py_ssize_t[2])
 */
struct __pyx_obj_18matrix_with_buffer_Matrix {
  PyObject_HEAD
  Py_ssize_t ncols;
  Py_ssize_t shape[2];
  Py_ssize_t strides[2];
  std::vector<float>  v;
};
```

[1] Yes, I know that in future python (PEP 563) is not not evaluating annotations during runtime, but I still think that cython should raise an error because we do evaluate annotations during compilation.

Reply:
> What I see as a pain in pure python mode is that it does not raise error that Py_ssize_t is not declared but it just fallbacks to object instead

I'm not sure about this. There's a lot of annotations that we don't understand (e.g. much of the `typing` module, pretty much any non-extension type class). The two things I could see about `Py_ssize_t` that could make it sensible to raise a warning are: it matches a C type, and Cython doesn't know where the name comes from (i.e. it isn't imported from anywhere). So we probably could do better about warning about annotations that Cython doesn't understand, but it shouldn't be all of them.

One of the reasons I proposed a "full" annotation typing mode (https://github.com/cython/cython/pull/4883) is that it could let us assume that every annotation is intended for Cython and actually generate errors if it doesn't understand them. I'm not sure my PR does that yet though.

> it is funny that `declare()` is fine with just Py_ssize_t

I think there was some debate about this too (https://github.com/cython/cython/issues/4882). My opinion was that `declare` is definitely intended to be read by Cython; an annotation might not be.

Reply:
Hmm my humble opinion is that the problem can be reduced to following: The code contains symbol (`Py_ssize_t`) which is not declared/imported anywhere - so cython should raise an error. Cython does not need to understand annotation but it should at least check if the symbol was declared somewhere or imported. It has nothing with understanding of annotations but "just" checking symbols properly. Cython can check symbols if are defined correctly and still ignore annotations since it cannot evaluate it properly.